### ISSUE TITLE
Drupal php error "Call to a member function inTransaction() on null ", php 8.3

### ISSUE THREAD

**Issue by jrglasgow** [ddev/ddev#5799]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Output of `ddev debug test`

<details><summary>Expand `ddev debug test` diagnostic information</summary>

```
Running bash [-c /var/folders/bd/w600ldtx14z4fdjhmk_3dj1m0000gn/T/test_ddev.sh]
======= Existing project config =========
These config files were loaded for project tsad8: [/Users/james/PhpStormProjects/tsad8_gov_build/.ddev/config.yaml]
name: tsad8
type: drupal10
docroot: docroot
php_version: 8.2
webserver_type: apache-fpm
webimage: ddev/ddev-webserver:v1.22.6
router_http_port: 80
router_https_port: 443
xdebug_enabled: true
additional_hostnames: []
additional_fqdns: []
database: {mariadb 10.4}
hooks: map[post-import-db:[map[exec:drush cr] map[exec:drush updb -y] map[exec:drush cr]] post-start:[map[composer:install --no-interaction]]]
upload_dir: docroot/sites/default/files
upload_dirs: [docroot/sites/default/files]
webimage_extra_packages: [exiftool]
project_tld: ddev.site
use_dns_when_possible: true
composer_version: 2
nodejs_version: 16
default_container_timeout: 120
======= Creating dummy project named  tryddevproject-7233 in ../tryddevproject-7233 =========
OS Information: Darwin Jamess-MacBook-Pro-2.local 22.3.0 Darwin Kernel Version 22.3.0: Mon Jan 30 20:42:11 PST 2023; root:xnu-8792.81.3~2/RELEASE_X86_64 x86_64
ProductName:		macOS
ProductVersion:		13.2.1
BuildVersion:		22D68
User information: uid=501(james) gid=20(staff) groups=20(staff),12(everyone),61(localaccounts),79(_appserverusr),80(admin),81(_appserveradm),98(_lpadmin),33(_appstore),100(_lpoperator),204(_developer),250(_analyticsusers),395(com.apple.access_ftp),398(com.apple.access_screensharing),399(com.apple.access_ssh),400(com.apple.access_remote_ae),701(com.apple.sharepoint.group.1)
DDEV version:  ITEM             VALUE
 DDEV version     v1.22.6
 architecture     amd64
 db               ddev/ddev-dbserver-mariadb-10.4:v1.22.6
 ddev-ssh-agent   ddev/ddev-ssh-agent:v1.22.6
 docker           24.0.7
 docker-compose   v2.23.3
 docker-platform  orbstack
 mutagen          0.17.2
 os               darwin
 router           ddev/ddev-traefik-router:v1.22.6
 web              ddev/ddev-webserver:v1.22.6
PROXY settings: HTTP_PROXY='' HTTPS_PROXY='' http_proxy='' NO_PROXY=''
======= DDEV global info =========
Global configuration:
instrumentation-opt-in=true
omit-containers=[]
performance-mode=mutagen
router-bind-all-interfaces=false
internet-detection-timeout-ms=3000
disable-http2=false
use-letsencrypt=false
letsencrypt-email=
table-style=bright
simple-formatting=false
use-hardened-images=false
fail-on-hook-fail=false
required-docker-compose-version=v2.23.3
use-docker-compose-from-path=false
project-tld=
xdebug-ide-location=
no-bind-mounts=false
router=traefik
wsl2-no-windows-hosts-mgt=false
router-http-port=80
router-https-port=443
mailpit-http-port=8025
mailpit-https-port=8026
traefik-monitor-port=10999

======= DOCKER info =========
docker location: lrwxr-xr-x  1 james  admin  34 Nov 22 12:27 /usr/local/bin/docker -> ../Cellar/docker/24.0.7/bin/docker
Docker Desktop Version: Print: Entry, ":CFBundleShortVersionString", Does Not Exist
Print: Entry, ":CFBundleVersion", Does Not Exist
Docker Desktop for Mac File build Doesn'tDocker Desktop for Mac Exist, build WillDocker Desktop for Mac Create: build /Applications/Docker.app/Contents/Info.plistDocker Desktop for Mac File build Doesn'tDocker Desktop for Mac Exist, build WillDocker Desktop for Mac Create: build /Applications/Docker.app/Contents/Info.plist
docker version:
Client: Docker Engine - Community
 Version:           24.0.7
 API version:       1.43
 Go version:        go1.21.3
 Git commit:        afdd53b4e3
 Built:             Thu Oct 26 07:06:42 2023
 OS/Arch:           darwin/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          24.0.7
  API version:      1.43 (minimum version 1.12)
  Go version:       go1.20.10
  Git commit:       311b9ff
  Built:            Thu Oct 26 09:05:28 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          v1.7.8
  GitCommit:        8e4b0bde866788eec76735cc77c4720144248fb7
 runc:
  Version:          1.1.9
  GitCommit:        82f18fe0e44a59034f3e1f45e475fa5636e539aa
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
DOCKER_DEFAULT_PLATFORM=notset
======= Mutagen Info =========
Mutagen is installed in ddev, version=0.17.2
--------------------------------------------------------------------------------
Name: tsad8
Identifier: sync_7CCGypSalx7q57scwLiaHn85edXexWcXOoMU3h9jvIk
Labels:
	com.ddev.config-hash: 654a7799e053320033f4baa45c0f9c345f2546e4
	com.ddev.volume-signature: Users-james-orbstack-run-1707336854
Configuration:
	Synchronization mode: Two Way Resolved
	Hashing algorithm: Default (SHA-1)
	Maximum allowed entry count: Default (2⁶⁴−1)
	Maximum staging file size: Default (18 EB)
	Symbolic link mode: POSIX Raw
	Ignore VCS mode: Default (Propagate)
	Ignores:
		/.git
		/.tarballs
		/.ddev/db_snapshots
		/.ddev/.importdb*
		.DS_Store
		.idea
		/docroot/sites/default/files
	Permissions mode: Default (Portable)
Alpha:
	URL: /Users/james/PhpStormProjects/tsad8_gov_build
	Configuration:
		Watch mode: Default (Portable)
		Watch polling interval: Default (10 seconds)
		Probe mode: Default (Probe)
		Scan mode: Default (Accelerated)
		Stage mode: Neighboring
		File mode: Default (0600)
		Directory mode: Default (0700)
		Default file/directory owner: Default
		Default file/directory group: Default
	Connected: No
Beta:
	URL: docker://ddev-tsad8-web/var/www/html
		DOCKER_HOST=unix:///Users/james/.orbstack/run/docker.sock
	Configuration:
		Watch mode: Default (Portable)
		Watch polling interval: Default (10 seconds)
		Probe mode: Default (Probe)
		Scan mode: Default (Accelerated)
		Stage mode: Neighboring
		File mode: Default (0600)
		Directory mode: Default (0700)
		Default file/directory owner: Default
		Default file/directory group: Default
		Compression: Default (DEFLATE)
	Connected: No
Status: [Paused]
--------------------------------------------------------------------------------
Name: tryddevproject-1452
Identifier: sync_XfiXzMUJDrKqXNTbaxuK9dKZ3PhmvYdbEMatq1PmNTa
Labels:
	com.ddev.config-hash: 0311359b0f6eedb428c2679de60b5d021c2d717b
	com.ddev.volume-signature: Users-james-orbstack-run-1707337210
Configuration:
	Synchronization mode: Two Way Resolved
	Hashing algorithm: Default (SHA-1)
	Maximum allowed entry count: Default (2⁶⁴−1)
	Maximum staging file size: Default (18 EB)
	Symbolic link mode: POSIX Raw
	Ignore VCS mode: Default (Propagate)
	Ignores:
		/.git
		/.tarballs
		/.ddev/db_snapshots
		/.ddev/.importdb*
		.DS_Store
		.idea
	Permissions mode: Default (Portable)
Alpha:
	URL: /Users/james/PhpStormProjects/tryddevproject-1452
	Configuration:
		Watch mode: Default (Portable)
		Watch polling interval: Default (10 seconds)
		Probe mode: Default (Probe)
		Scan mode: Default (Accelerated)
		Stage mode: Neighboring
		File mode: Default (0600)
		Directory mode: Default (0700)
		Default file/directory owner: Default
		Default file/directory group: Default
	Connected: Yes
	Synchronizable contents:
		28 directories
		76 files (106 kB)
		0 symbolic links
Beta:
	URL: docker://ddev-tryddevproject-1452-web/var/www/html
		DOCKER_HOST=unix:///Users/james/.orbstack/run/docker.sock
	Configuration:
		Watch mode: Default (Portable)
		Watch polling interval: Default (10 seconds)
		Probe mode: Default (Probe)
		Scan mode: Default (Accelerated)
		Stage mode: Neighboring
		File mode: Default (0600)
		Directory mode: Default (0700)
		Default file/directory owner: Default
		Default file/directory group: Default
		Compression: Default (DEFLATE)
	Connected: Yes
	Synchronizable contents:
		28 directories
		76 files (106 kB)
		0 symbolic links
Status: Watching for changes
--------------------------------------------------------------------------------
======= Docker Info =========
Docker platform: orbstack
Using Docker context: default (unix:///Users/james/.orbstack/run/docker.sock)
docker-compose: v2.23.3
Using DOCKER_HOST=unix:///Users/james/.orbstack/run/docker.sock
Docker version: 24.0.7
Able to run simple container that mounts a volume.
Able to use internet inside container.
Docker disk space:
Filesystem                Size      Used Available Use% Mounted on
overlay                 162.7G     66.3G     96.4G  41% /

 Container ddev-tryddevproject-1452-db  Stopped
 Container ddev-tryddevproject-1452-web  Stopped
 Container ddev-tryddevproject-1452-web  Stopped
 Container ddev-tryddevproject-1452-db  Stopped
 Container ddev-tryddevproject-1452-db  Removed
 Container ddev-tryddevproject-1452-web  Removed
 Network ddev-tryddevproject-1452_default  Removed
Project tryddevproject-1452 has been stopped.
Stopped Mutagen daemon
The ddev-ssh-agent container has been removed. When you start it again you will have to use 'ddev auth ssh' to provide key authentication again.
Network ddev_default removed
Existing docker containers:
CONTAINER ID   IMAGE          COMMAND     CREATED      STATUS    PORTS     NAMES
6c750dfb9524   17b365050b12   "/bin/sh"   7 days ago   Created             phpstorm_helpers_PS-223.8836.42
Network ddev_default created
Starting tryddevproject-7233...
Network ddev-tryddevproject-7233_default created
 Container ddev-ssh-agent  Created
 Container ddev-ssh-agent  Started
ssh-agent container is running: If you want to add authentication to the ssh-agent container, run 'ddev auth ssh' to enable your keys.
Building project images...
Project images built in 0s.
 Container ddev-tryddevproject-7233-web  Created
 Container ddev-tryddevproject-7233-db  Created
 Container ddev-tryddevproject-7233-web  Started
 Container ddev-tryddevproject-7233-db  Started
You have Mutagen enabled and your 'php' project type doesn't have `upload_dirs` set.
For faster startup and less disk usage, set upload_dirs to where your user-generated files are stored.
If this is intended you can disable this warning with `ddev config --disable-upload-dirs-warning`.
Starting Mutagen sync process...
Mutagen sync flush completed in 1s.
For details on sync status 'ddev mutagen st tryddevproject-7233 -l'
Waiting for web/db containers to become ready: [web db]
Starting ddev-router if necessary...
 Container ddev-router  Created
 Container ddev-router  Started
Waiting for additional project containers to become ready...
All project containers are now ready.
Successfully started tryddevproject-7233
Project can be reached at https://tryddevproject-7233.ddev.site https://127.0.0.1:32852
======== Curl of site from inside container:
HTTP/1.1 200 OK
Server: nginx
Date: Wed, 07 Feb 2024 20:23:37 GMT
Content-Type: text/html; charset=UTF-8
Connection: keep-alive
Vary: Accept-Encoding

======== curl -I of http://tryddevproject-7233.ddev.site from outside:
HTTP/1.1 200 OK
Content-Type: text/html; charset=UTF-8
Date: Wed, 07 Feb 2024 20:23:37 GMT
Server: nginx
Vary: Accept-Encoding

======== full curl of http://tryddevproject-7233.ddev.site from outside:
Success accessing database... db via TCP/IP
ddev is working. You will want to delete this project with 'ddev delete -Oy tryddevproject-7233'
======== Project ownership on host:
drwxr-xr-x  4 james  staff  128 Feb  7 13:22 ../tryddevproject-7233
======== Project ownership in container:
drwxr-xr-x 1 james dialout 16 Feb  7 20:23 /var/www/html
======== In-container filesystem:
Filesystem     Type  1K-blocks     Used Available Use% Mounted on
/dev/vdb1      btrfs 170573192 69685296 100946440  41% /var/www
======== curl again of tryddevproject-7233 from host:
Success accessing database... db via TCP/IP
ddev is working. You will want to delete this project with 'ddev delete -Oy tryddevproject-7233'
Thanks for running the diagnostic. It was successful.
Please provide the output of this script in a new gist at gist.github.com
Running ddev launch in 5 seconds

Please run cleanup after debugging with 'ddev debug testcleanup'
```
</details>         

### Expected Behavior

no database issues

### Actual Behavior

PHP loses connection to MySQL during transactions. This occurs when running som Drupal update hooks, 

```
PHP Fatal error:  Uncaught Error: Call to a member function inTransaction() on null in /var/www/html/docroot/core/modules/mysql/src/Driver/Database/mysql/Connection.php:483
Stack trace:
#0 /var/www/html/docroot/core/modules/mysql/src/Driver/Database/mysql/Connection.php(415): Drupal\mysql\Driver\Database\mysql\Connection->doCommit()
#1 /var/www/html/docroot/core/lib/Drupal/Core/Database/Connection.php(1447): Drupal\mysql\Driver\Database\mysql\Connection->popCommittableTransactions()
#2 /var/www/html/docroot/core/lib/Drupal/Core/Database/Transaction.php(71): Drupal\Core\Database\Connection->popTransaction('drupal_transact...')
#3 [internal function]: Drupal\Core\Database\Transaction->__destruct()
#4 {main}
  thrown in /var/www/html/docroot/core/modules/mysql/src/Driver/Database/mysql/Connection.php on line 483

Fatal error: Uncaught Error: Call to a member function inTransaction() on null in /var/www/html/docroot/core/modules/mysql/src/Driver/Database/mysql/Connection.php on line 483

Error: Call to a member function inTransaction() on null in /var/www/html/docroot/core/modules/mysql/src/Driver/Database/mysql/Connection.php on line 483

Call Stack:
    9.9992   78603384   1. Drupal\Core\Database\Transaction->__destruct() /var/www/html/docroot/core/lib/Drupal/Core/Database/Transaction.php:0
    9.9992   78603384   2. Drupal\Core\Database\Connection->popTransaction($name = 'drupal_transaction') /var/www/html/docroot/core/lib/Drupal/Core/Database/Transaction.php:71
    9.9992   78603384   3. Drupal\mysql\Driver\Database\mysql\Connection->popCommittableTransactions() /var/www/html/docroot/core/lib/Drupal/Core/Database/Connection.php:1447
    9.9992   78603760   4. Drupal\mysql\Driver\Database\mysql\Connection->doCommit() /var/www/html/docroot/core/modules/mysql/src/Driver/Database/mysql/Connection.php:415
```

### Steps To Reproduce

As best as I can determine problem is that sometime after the database connection is made the Connection variable gets set to NULL.

I upgraded to DDEV v1.22.5 to get access to PHP 8.3, then to v1.22.6 to get access to xdebug in PHP 8.3. I began to have problems in December seeing the "call to member function on null" error. I spent some time struggling with it before a co-worker commented about having problems. Comparing notes we determined that we were having the same problem, talked to a third co-worker and when he wasn't having problem working on the same project determined that the DDEV version was the issue. Downgrading back to v1.22.5 solved the problem, but I lose access to xdebug in PHP 8.3 ☹️.
I am experiencing this problem with at least two different websites and ddev configurations (granted one DDEV config is a copy of the other). Different code bases (both Drupal 10) and databases. 

I would think it was a Drupal 10 or a module issue if it weren't for the fact that downgrading DDEV fixes the problem.

**Comment by rfay** [comment]
This sure looks like a problem with your PHP code. I don't think you can get this error without a problem in your PHP code.

I'm sure you know that different versions of DDEV have different patch version of PHP 8.3. What do you have? 

You should be able to use xdebug just fine with `webimage_extra_packages: [php8.3-xdebug]` which gets you the latest xdebug version. Also, DDEV v1.22.7 will be out today or tomorrow with an updated image.

**Comment by rfay** [comment]
DDEV v1.22.7 is now available, which has PHP 8.3.2 and Xdebug 3.3.1. Could you please test with it?

What you're seeing is almost certainly a code problem, but could have something to do with PHP patch version I supposed.

**Comment by jrglasgow** [comment]
I have verified that the problem still exists for me in DDEV v1.22.7... I will now begin comparing projects and try to see which code is shared between them all that might be causing the problem.

**Comment by jrglasgow** [comment]
In my investigating I found this Drupal.org issue [Transaction autocommit during shutdown relies on unreliable object destruction order (xdebug 3.3+ enabled)](https://www.drupal.org/project/drupal/issues/3405976) which makes it appear to be an xdebug change that caused the problem. This is why changing the PHP version from 8.3 to 8.2 or 8.2 didn't fix the issue, only rolling back to ddev v1.22.5... I assume that DDEV v1.22.5 has a different xdebug version,

**Comment by rfay** [comment]
Yes, thanks. that issue is actually mentioned in the [v1.22.7 release notes](https://github.com/ddev/ddev/releases/tag/v1.22.7) and has been widely shared as well. Thanks for reporting back. 

I believe xdebug for PHP 8.3 wasn't available until 1.22.6, and that it became available because of xdebug 3.3.*

**Comment by pasan-dg** [comment]
Hi, I'm using the following configuration with Php 8.1 and was using ddev 1.22.7 but had same error when enabling xdebug.
I've updated ddev to 1.23.0 but I still get the same error.
What would be the solution for this please ?
Thanks

name: project
type: drupal9
docroot: public
php_version: "8.1"
webserver_type: nginx-fpm
xdebug_enabled: false
additional_hostnames: []
additional_fqdns: []
database:
    type: mysql
    version: "8.0"
use_dns_when_possible: true
composer_version: "2"
web_environment: []

**Comment by rfay** [comment]
Xdebug 3.3.2 has been released and claims to solve some of the crashes. It is in the upstream deb.sury.org packages as of yesterday. You may get relief with
```
webimage_extra_packages: ["php${DDEV_PHP_VERSION}-xdebug"]
```

Do a `ddev restart` to pick it up.

Please report back on your results

**Comment by pasan-dg** [comment]
Hi @rfay 
Thank you for the fast response.
I tried adding the line and ddev restart
But have same error.

```PHP Fatal error:  Uncaught Error: Call to a member function inTransaction() on null in /var/www/html/public/core/modules/mysql/src/Driver/Database/mysql/Connection.php:470
Stack trace:
#0 /var/www/html/public/core/modules/mysql/src/Driver/Database/mysql/Connection.php(402): Drupal\mysql\Driver\Database\mysql\Connection->doCommit()
#1 /var/www/html/public/core/lib/Drupal/Core/Database/Connection.php(1678): Drupal\mysql\Driver\Database\mysql\Connection->popCommittableTransactions()
#2 /var/www/html/public/core/lib/Drupal/Core/Database/Transaction.php(71): Drupal\Core\Database\Connection->popTransaction('drupal_transact...')
#3 [internal function]: Drupal\Core\Database\Transaction->__destruct()
#4 {main}
  thrown in /var/www/html/public/core/modules/mysql/src/Driver/Database/mysql/Connection.php on line 470

Fatal error: Uncaught Error: Call to a member function inTransaction() on null in /var/www/html/public/core/modules/mysql/src/Driver/Database/mysql/Connection.php on line 470

Error: Call to a member function inTransaction() on null in /var/www/html/public/core/modules/mysql/src/Driver/Database/mysql/Connection.php on line 470

Call Stack:
    5.0119   61282616   1. Drupal\Core\Database\Transaction->__destruct() /var/www/html/public/core/lib/Drupal/Core/Database/Transaction.php:0
    5.0119   61282616   2. Drupal\Core\Database\Connection->popTransaction($name = 'drupal_transaction') /var/www/html/public/core/lib/Drupal/Core/Database/Transaction.php:71
    5.0119   61282616   3. Drupal\mysql\Driver\Database\mysql\Connection->popCommittableTransactions() /var/www/html/public/core/lib/Drupal/Core/Database/Connection.php:1678
    5.0119   61282992   4. Drupal\mysql\Driver\Database\mysql\Connection->doCommit() /var/www/html/public/core/modules/mysql/src/Driver/Database/mysql/Connection.php:402
```

```ddev php --version
PHP 8.1.27 (cli) (built: Apr 11 2024 22:37:31) (NTS)
Copyright (c) The PHP Group
Zend Engine v4.1.27, Copyright (c) Zend Technologies
    with Zend OPcache v8.1.27, Copyright (c), by Zend Technologies
    with Xdebug v3.3.2, Copyright (c) 2002-2024, by Derick Rethans
```

**Comment by pasan-dg** [comment]
I'm currently restricted to use 9.5.11

**Comment by pasan-dg** [comment]
Thanks for the help @rfay

**Comment by rfay** [comment]
I was unaware that the  issue affected Drupal 9, which is EOL anyway. So you definitely won't get the fix. 

I'm no expert on the problem, but it appears that it was discovered in 10.2.x 5 months ago. 

You should be able to use the workarounds in https://stackoverflow.com/questions/77582344/how-can-i-install-xdebug-on-ddev-for-php-8-3-before-its-ready-in-a-release/77583024#77583024 to downgrade xdebug to 3.2.x as mentioned in the Drupal.org issue.

The maintainer of Xdebug believes that these issues are resolved in 3.3.2, https://github.com/xdebug/xdebug/releases/tag/3.3.2
